jane/doc/extensions/uniqueness/reference.md

```diff
@@ -0,0 +1,103 @@
+# Uniqueness Reference
+
+The goal of this document is to be a reasonably complete reference to the
+uniqueness mode. For a gentler introduction, see [the introduction](intro.md).
+It will grow over time as we add more features such as overwriting and
+borrowing.
+
+## More details on uniqueness checking
+
+### Mixing unique and aliased uses
+
+Continuing the example from [the introduction](intro.md), we can have both a
+unique and an aliased access to a value, as long as those occur in separate
+branches:
+
+```ocaml
+val free : t @ unique -> unit
+val store : t @ aliased -> unit
+
+let okay t =
+  if flip_coin ()
+  then free t
+  else (store t; store t)
+```
+
+This might be surprising coming from a language that uses ownership like Rust.
+In Rust, each value needs to have exactly one owner which is responsible for its
+deallocation. But in OCaml, we can rely on the GC for deallocation which makes
+it possible to give up ownership of the value in the second branch.
+
+### Unique uses of different fields
+
+The uniqueness analysis makes it possible to consume some fields of an
+allocation without losing access to the allocation or its other fields. For
+example, you might write:
+
+```ocaml
+let okay r =
+  free r.field1;
+  match r with
+  | { field2; _ } -> free field2
+```
+
+This is fine: we free the first field of `r`, then pattern match on `r` itself
+(which was not affected by the free of its child) and then free the second
+child. However, it would not be okay to free the first field and then free `r`
+itself (which has now partially been consumed):
```

Comment:
I wonder if instead of `free` which has certain connotation, we should just say `use_uniquely` which is more generic. For the current example, the reader might think: well I've freed `r.field1`, but I can still free the memory block of `r` itself (which doesn't contain `r.field1`).

Reply:
Right, it is a bit confusing because in C `free` only applies to the current block rather than the whole structure. Whereas here it applies to the children as well. I'll rephrase to make that clearer.

`store` is just some function that takes an `aliased` argument (perhaps to store it in a global reference somewhere). Let me know if you prefer a different name.

Reply:
Instead of `store`, maybe just `use_as_aliased`? I'm also fine with `store` as long as it's explained.